Stratis.Bitcoin/BlockPulling/LookaheadBlockPuller.cs

```diff
@@ -10,299 +10,400 @@
+        /// </summary>
+        /// <param name="count">How many blocks ahead (minus one) should the returned block be ahead of the current puller's position. 
+        /// A value of zero will provide the next block, a value of one will provide a block that is 2 blocks ahead.</param>
+        /// <returns>The block which height is <paramref name="count"/>+1 higher than current puller's position, 
+        /// or null if such a block is not downloaded or does not exist.</returns>
+        Block TryGetLookahead(int count);
+
+        /// <summary>Sets the current location of the puller to a specific block header.</summary>
+        /// <param name="location">Block header to set the location to.</param>
+        void SetLocation(ChainedBlock location);
+
+        /// <summary>
+        /// Waits for a next block to be available (downloaded) and returns it to the consumer.
+        /// </summary>
+        /// <param name="cancellationToken">Cancellation token to allow the caller to cancel waiting for the next block.</param>
+        /// <returns>Next block or null if a reorganization happened on the chain.</returns>
+        Block NextBlock(CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Adds a specific requirement to all peer nodes.
+        /// </summary>
+        /// <param name="transactionOptions">Specifies the requirement on nodes to add.</param>
+        void RequestOptions(TransactionOptions transactionOptions);
+    }
+
+    /// <summary>
+    /// Puller that is used for fast sync during initial block download (IBD). It implements a strategy of downloading 
+    /// multiple blocks from multiple peers at once, so that IBD is fast enough, but does not consume too many resources.
+    /// </summary>
+    /// <remarks>
+    /// The node is aware of the longest chain of block headers, which is stored in this.Chain. This is the chain the puller 
+    /// needs to download. Here is a visualization of the block chain and how the puller sees it:
+    /// <para>
+    /// -------A------B-------------C-----------D---------E--------
+    /// </para>
+    /// <para>
+    /// Each '-' represents a block and letters 'A' to 'E' represent blocks with important positions in the chain from 
+    /// the puller's perspective. The puller can be understood as a producer of the blocks (by requesting them from the peers 
+    /// and downloading them) for the component that uses the puller that consumes the blocks (e.g. validating them).
+    /// </para>
+    /// <para>
+    /// A is a position of a block that we call location. Blocks in front of A were already downloaded and consumed.
+    /// </para>
+    /// <para>
+    /// B is a position of a block A + MinimumLookahead, and E is a position of block A + MaximumLookahead. 
+    /// Blocks between B and E are the blocks that the puller is currently interested in. Blocks after E are currently 
```

Comment:
B and E are here only to make sure that ActualLookAhead has a sane value.

[1,  1 + ActualLookahead] is the currently downloading blocks